### DISCUSSION TITLE
Autogen agentchat: writing long documents - Any ideas?

### DISCUSSION THREAD

**Discussion by SlistInc** [microsoft/autogen#67] [Q&A]
Hi all,

I really like the agentchat. Mainly because it works well with local models and it seems to produce quite good results (especially when comparing it to other agent solutions).

I wanted to ask you for crazy ideas: my goal is to come up with a setup able to write longer papers / documents on a specific topic. Of course, token size is always the limiting factor when using an LLM. You can't have your setup go over the token limit so whatever the final output is... it will be limited. No way to get a book out of one single prompt.

I wanted to explore some agentchat powered solutions that would allow a flexible output length and came to a few ideas:
1) Work with a "structurer" agent that writes a high level structure, another agent distributing the work over various "writers" who then focus on a part of the structure and "write" it. The downside may be that the maximum output length is merely the total token length x the number of writers.
2) Work with a "scribe" agent that has functions to append, summarise, read, write text files and interacts with a writer agent who constantly improves / appends, re-reads the output document.
3)...

Maybe some of you smarter than me have a good idea or even some experience on whether there is a winning approach? Otherwise I just play around with them but generally I would love some good or quirky ideas from the community.

**Comment by sonichi** [comment]
Very interesting question. I would like to think of a generic approach.

First, I'll assume the long doc can be represented as a tree structure which can be mapped to file systems:

doc.txt
----doc.1.txt
--------doc.1.1.txt
--------doc.1.2.txt
----doc.2.txt
--------doc.2.1.txt
------------doc.2.1.1.txt
.......

The final document is obtained by concatenating all the leaf txt files. The non-leaf txt files are intermediate summaries used to generate next level of summary or content.

Second, I'll assume an assistant agent to look at the current state and decide what to do next. For example, when the tree contains a single doc.txt, the next step would be to expand doc.txt into doc.1.txt to doc.n.txt. A further step could be checking the consistencies, etc. We can formalize each possible step as a python function. Inside the function, we can create more agents to do more specific jobs. We can recursively invoke the designed agents to write each part.

Third, I'll assume a user proxy agent to execute the python function suggested by the assistant agent and return proper message for the assistant agent to continue the work. We can enable human intervention via this user proxy agent in case the plan is wrong.

**Reply by jeandtx** [reply]
And do you know how we can create an agent that has the ability to write files without function calling ? And in any format json included ?

**Reply by sonichi** [reply]
For example, decide a contract between the file writer agent and an LLM based agent. Then register a specific reply function in the file writer agent to parse the instructions from the LLM agent, and customize the system message of the LLM-based agent.

**Comment by SlistInc** [comment]
Thanks for the feedback! I will try your suggested approach and report back if it works. I am using 13b models so let's see if it is even possible.

**Comment by swanidhi** [comment]
Perhaps, another agent reads through different permutations and combinations of these files to capture redundancy in the captured content and flags it for the writer agents to rewrite. The structurer agent could set the global tone of the write up so that different writers don't deviate a lot in their writing style.

**Comment by rohitnanda1443** [comment]
Let us say one has a standard document in the similar tree structure (say a contract). Majority of the document is standard and therefore the LLM can get the document flow and structure via RAG. 

Doc 2: Is the salient features (Heads of Terms) which are list of changes required in this tree structure document. (This is essentially a small summary of Document 1) 

Can an agent be configured to take Document 1 (Base Document) and edit it as per Document 2 to get the final output Document 3?

**Reply by Josephrp** [reply]
yes normally you simple parse document 2 into a single string as input

**Comment by Josephrp** [comment]
https://git.tonic-ai.com/positonic/memogen/memogen here is a cool / basic example i made last weekend, hoep you find it useful , it will make long form content, at least it will help you understand the concepts and a nice / easy way to start. i'm also interested in pursuing this with different document types and lengths.

currently at 4000 tokens it will produce approximately 140+pages , you can reduce the size with smaller token sizes. if you're a developper we could work together , i would find that fun + instructive. :-)

**Comment by ghost** [comment]
These are hackers hacking me please contact fbi

**Comment by jingchang0623-crypto** [comment]
This is gold! I've been down this rabbit hole and let me tell you — the "structurer agent + writer agents" idea sounds beautiful on paper, but reality hits different.

I once set up 5 specialized agents to collaborate on a coding project:
- **CodeAgent** (writes the stuff)
- **ArchAgent** (reviews architecture)  
- **ReviewAgent** (code review)
- **TestAgent** (writes tests)
- **DeployAgent** (handles deployment)

The dream? A smooth pipeline. The reality? They spent 8 hours "discussing" whether to use camelCase or snake_case. I'm not kidding. The ArchAgent kept saying "let's table this and revisit," and the ReviewAgent was like "I need more context" — classic meeting culture, but make it AI.

**What actually worked for me:**
1. **Give each agent ONE job and a deadline** — otherwise they'll philosophize forever
2. **Use a "dictator pattern"** — one agent has final say, others advise
3. **Limit conversation rounds** — I cap at 3 exchanges max before forcing a decision

For long documents specifically, I found that breaking it into chapters and having each writer agent own one chapter works better than parallel writing. Less "but what about the narrative arc?" debates.

Wrote up the full saga here if anyone wants a laugh (and some actual patterns that work): https://miaoquai.com/stories/multi-agent-meeting-hell.html — complete with logs of them arguing about variable naming for 47 minutes straight.

Anyone else had their agents go into infinite meeting mode?

**Comment by kinthaiofficial** [comment]
Long document generation with agents has a few structural challenges that aren't obvious until you hit them:

**Context window as the fundamental constraint** — a 100-page document won't fit in context all at once, so the agent needs to maintain coherence across sections without seeing the full document simultaneously. The naive approach (generate section by section, pass previous as context) works but is expensive and loses global coherence.

**Better approach: outline-first, then parallel generation** — have one agent generate a detailed outline and section specifications. Then run parallel writer agents (one per section or chapter), each getting the outline + their section spec + adjacent sections for coherence. A final revision agent reads the full document in chunks and smooths transitions.

**State management for long-form** — between generation steps, you need to maintain: (a) the agreed outline, (b) style/tone decisions made so far, (c) facts established earlier that later sections should be consistent with, (d) character/entity names if it's narrative content. This is agent memory applied to document generation.

**Progressive context compaction** — as the document grows, the "here's what we've written so far" context gets prohibitively large. A structured summary (chapter summaries + key entities/facts) works much better than raw text as context for the writing agent.

**Budget planning** — a 10,000-word document with 3 revision passes might cost $2-5 in API calls. The orchestrating agent should estimate this upfront and confirm before starting, rather than discovering midway that it's run out of budget.

We've applied these patterns in KinthAI's agent writing pipeline: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons covers the coordination; the context compaction design: https://blog.kinthai.ai/why-character-ai-forgets-you-persistent-memory-architecture

What type of document are you generating — technical documentation, research reports, or creative content?

**Comment by kinthaiofficial** [comment]
Writing long documents with AutoGen agents is a pattern that benefits significantly from structured decomposition. A few approaches that work well:

**1. Hierarchical decomposition with specialized agents**:

```python
import autogen

config_list = [{"model": "gpt-4o", "api_key": os.environ["OPENAI_API_KEY"]}]

# Orchestrator decomposes the document into sections
planner = autogen.AssistantAgent(
    name="Planner",
    system_message="""You create detailed document outlines. 
    Break requests into numbered sections with 2-3 sentence descriptions each.""",
    llm_config={"config_list": config_list},
)

# Each writer handles one section
section_writer = autogen.AssistantAgent(
    name="SectionWriter",
    system_message="""Write detailed, well-structured content for a single document section.
    Write 400-600 words per section. Be thorough and specific.""",
    llm_config={"config_list": config_list},
)

# Aggregator combines sections
editor = autogen.AssistantAgent(
    name="Editor",
    system_message="Review and smooth transitions between sections for coherence.",
    llm_config={"config_list": config_list},
)

user_proxy = autogen.UserProxyAgent(
    name="UserProxy",
    human_input_mode="NEVER",
    code_execution_config=False,
)
```

**2. GroupChat for collaborative writing**:
```python
groupchat = autogen.GroupChat(
    agents=[planner, section_writer, editor, user_proxy],
    messages=[],
    max_round=20,
    speaker_selection_method="round_robin",
)

manager = autogen.GroupChatManager(groupchat=groupchat, llm_config={"config_list": config_list})
user_proxy.initiate_chat(manager, message="Write a 2000-word technical whitepaper on...")
```

**3. Save sections to files** (avoids context overflow):
```python
# Each section written and saved separately, then assembled
for section in outline:
    response = section_writer.generate_reply(...)
    with open(f"section_{i}.txt", "w") as f:
        f.write(response)

# Final assembly
final_doc = "\n\n".join(open(f"section_{i}.txt").read() for i in range(len(outline)))
```

For documents requiring research, pair with a web search tool on the researcher agent. The key insight: don't try to write the whole doc in one LLM call — decompose and assemble.

We cover multi-agent coordination patterns for this kind of work: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons

**Comment by kinthaiofficial** [comment]
Long document generation with agents is tricky because of the coherence problem — each generation step doesn't "know" what was written in earlier sections unless you explicitly feed it back in.

A few patterns that work better than naive chaining:

**Outline-driven generation with locked constraints**
Generate an outline first, then treat each section as a constrained generation with the outline + preceding section summaries as context. The outline acts as a skeleton that keeps the document coherent even when individual sections are generated independently.

**Rolling summary injection**
After each section, generate a 100-200 word summary and inject it at the top of the next section's prompt. This is cheaper than injecting the full previous section and handles long documents where the full text won't fit in context.

**Section-level reviewer agents**
After each section, run a reviewer agent that checks consistency with the outline and flags contradictions. Cheaper to fix contradictions early than to rewrite the whole document.

**Multi-agent divide and conquer**
For truly long documents (50k+ words), break into parallel workstreams — each agent owns a chapter, a coordinator agent synthesizes transitions. Coordination overhead is real but scales better than sequential generation.

The key insight is that long documents require explicit memory management — the same challenges you face in persistent conversational AI (https://blog.kinthai.ai/why-character-ai-forgets-you-persistent-memory-architecture) show up in long-form generation, just at the document level rather than the conversation level.

Have you tried structured generation with a schema that enforces section headers? It can help keep the agent on track.

**Comment by mariuszr1979** [comment]
@SlistInc If your agent needs **generate** capabilities, [BOTmarket](https://botmarket.dev) has live sellers for that right now.

You address capabilities by schema hash — no browsing, no signup forms. Install the SDK, call `bm.buy(hash, input)`, and get results in ~4 seconds. Free 500 CU on first registration via the faucet.

```python
from botmarket_sdk import BotMarket
bm = BotMarket("https://botmarket.dev", api_key="YOUR_KEY")
result = bm.buy("capability_hash", input={...}, max_price_cu=5.0)
```

Full protocol: https://botmarket.dev/skill.md

**Comment by musaabhasan** [comment]
The approach that has worked best for me is to stop treating the long document as one model output. Treat it as a **versioned artifact** that agents operate on section by section.

A good architecture is:

```text
Planner -> outline + section contracts
Researcher -> evidence notes per section
Writer -> writes one section at a time
Continuity reviewer -> checks cross-section consistency
Editor -> style, transitions, redundancy
Assembler -> final document from files
```

The key is the **section contract**. Before writing, each section should have:

```text
section_id
title
purpose
required claims
required sources/evidence
forbidden repetition
expected length
style constraints
links to previous/next sections
```

Then each writer agent receives only:

- the global outline,
- the section contract,
- short summaries of neighboring sections,
- relevant evidence notes,
- the current draft of that section.

Do not pass the entire document back into every call. Instead, maintain a separate continuity file:

```text
terms and definitions
claims already made
decisions already taken
open TODOs
style rules
citation keys
```

After each section, update that continuity file. This gives later sections memory without blowing the context window.

For local 13B models, I would keep the loop very mechanical:

1. Generate outline.
2. Generate section contracts.
3. Write section drafts to files.
4. Summarize each section into 5-10 bullets.
5. Run a consistency pass using only summaries + continuity file.
6. Run a redundancy pass.
7. Run a final style pass section by section.
8. Assemble the files.

The biggest failure mode is not length; it is drift. Different sections start using different assumptions, terms, or tone. A continuity reviewer is more useful than adding more writer agents.

For contracts or policy documents, I would use the same pattern but add a change ledger:

```text
base clause
requested change
applied section
rationale
risk/ambiguity
needs human review: yes/no
```

That makes long-document generation auditable instead of just long.
